android-components/components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt

```diff
@@ -93,6 +99,23 @@ abstract class PlacesStorage(
         readScope.coroutineContext.cancelChildren()
     }
 
+    /**
+     * Cleans up pending read operations of a specific query.
+     *
+     * @param queryToCancel Previous query to cancel reads for.
+     * Calling cancel multiple times for the same query has effect only the first time.
+     * Use this in scenarios where the same instance is used in multiple scenarios to prevent cases
+     * in which a general cancel operation for one scenario cancels other reads for the same query.
+     * If the value is an empty string all current reads are immediately cancelled.
+     */
+    override fun cancelReadsOfQuery(queryToCancel: String) {
+        if (queryToCancel.isEmpty() || lastCancelledQuery != queryToCancel) {
+            lastCancelledQuery = queryToCancel
```

Comment:
❓ I thought `lastCancelledQuery` is not shared between storage? (ex: bookmarksStorage, historyStorage and historyMetadataStorage don't have the same `lastCancelledQuery`)

Reply:
Good catch, tested now again.
Indeed Fenix is currently using two different instances of `PlacesStorage`, one for [bookmarks](https://searchfox.org/mozilla-mobile/rev/7c70c3df303e19acc69877821c3770fe85393005/fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt#357) and one for [history](https://searchfox.org/mozilla-mobile/rev/7c70c3df303e19acc69877821c3770fe85393005/fenix/app/src/main/java/org/mozilla/fenix/components/Core.kt#356).
So between the providers of this data there should be no overlapping.

But on Fenix the `CombinedHistoryProvider` will use [the same `PlacesStorage` instance](https://searchfox.org/mozilla-mobile/rev/7c70c3df303e19acc69877821c3770fe85393005/fenix/app/src/main/java/org/mozilla/fenix/search/awesomebar/AwesomeBarView.kt#126-127) for both metadata and "legacy" history.
And then I planned to use this new provider with the same instance also - as the instance is quite big and we're interested in only a very small part of it's functionality.
This will mean that we need to somehow prevent overlapping cancel requests which is why I've added this new `cancelReads(nextQuery: String)` api.
(Using the same storage instance for multiple queries is no less performant than using different storage instances for multiple queries as all calls will query the same database from application-services so they would be executed serially either way.)
So I found it simpler, probably better for performance - memory and safer for scenarios in which the same underlying storage is used for multiple providers - prevent unintentional overlapping cancel requests.